File: packages/shared-types/src/history.ts
export type ProblemSource = "lesson" | "generated" | string;

export type AttemptSummary = {
  id: string;
  problem_source: ProblemSource;
  lesson_id: string | null;
  generated_problem_id: string | null;
  lesson_title: string | null;
  is_correct: boolean;
  score: number;
  attempted_at: string;
};

export type AttemptDetail = AttemptSummary & {
  submitted_code: string | null;
  answer: Record<string, unknown> | null;
  ai_review: string | null;
  time_spent_seconds: number | null;
  problem_content: Record<string, unknown> | null;
};

export type HistoryListResponse = {
  items: AttemptSummary[];
  next_cursor: string | null;
};

File: apps/backend/app/schemas/history.py
from datetime import datetime

from pydantic import BaseModel


class AttemptSummary(BaseModel):
    id: str
    problem_source: str
    lesson_id: str | None
    generated_problem_id: str | None
    lesson_title: str | None = None
    is_correct: bool
    score: int
    attempted_at: datetime

    model_config = {"from_attributes": True}


class AttemptDetail(AttemptSummary):
    submitted_code: str | None
    answer: dict | None
    ai_review: str | None
    time_spent_seconds: int | None
    problem_content: dict | None = None


class HistoryListResponse(BaseModel):
    items: list[AttemptSummary]
    next_cursor: str | None = None
